### DISCUSSION TITLE
linux compatibility

### DISCUSSION THREAD

**Discussion by knkitten** [SteamRE/SteamKit#1056] [General]
Can I use NetHook2 in some linux distributions (for ex. ubuntu)?

Let's say I successfully compile the project (got NetHook2.dll? or smth similar). 
But can I now inject NetHook2.dll similar to how it is done in Windows: `rundll32 "<Path To NetHook2.dll>",Inject` ? 

There may be analogs of `rundll32` for linux?

**Comment by yaakov-h** [comment]
NetHook2 relies on MS Detours and is very specific to Windows, and I think even 32-bit Steam at that.

On Linux and macOS we don't necessarily need to self-inject and should be able to rely on `LD_PRELOAD` and `DYLD_INSERT_LIBRARIES` respectively, but we would still need a way to identify the Steam memory locations containing the methods we want to override and be able to override them.

I'm not sure how feasible that is, it's something I looked into quite a few years ago but couldn't find any answers for.

**Reply by JustArchi** [reply]
If by any chance you're looking for additional encouragement then I'd be very interested in having working nethook2 for linux as well.

But as I know how much of potential work that is, I don't count on it happening anytime soon, if at all - more of a wishlist thingy for the future :smiling_face_with_tear:.

**Reply by yaakov-h** [reply]
If you can figure out how to make it happen, then be my guest. :)

**Reply by Nemirtingas** [reply]
Hi, is any one of you still working on this ? I might be able to help you there, hooking stuff on linux.

**Reply by yaakov-h** [reply]
Not actively. @m4dEngi did it in #1082 for some April Fools fun but I haven't gone back to look at integrating it properly.

**Reply by Nemirtingas** [reply]
What function do I need to hook, how I can find them to test ? My code should be ready. It compiles, I can inject it into steam process but obviously the code signature is different than Windows.
I got the new pattern for Linux, nvm, I found my way.

**Reply by Nemirtingas** [reply]
I made progress, I successfully dumped input message from a linux steam exe. My branch is here: https://github.com/Nemirtingas/SteamKit/tree/linux_hook
I've populated the read with minimal how to.

**Reply by m4dEngi** [reply]
> Not actively. @m4dEngi did it in #1082 for some April Fools fun but I haven't gone back to look at integrating it properly.

That was a pretty messed up attempt ( yepp, just wanted to make tux hook pun ). Think i figured out how to do it properly since then, but i don't think it can be integrated the way you want it. There's too much platform specific code to just lump it all together and not making a mess, or which is more likely, i'm just bad at it. 


> What function do I need to hook, how I can find them to test ? My code should be ready. It compiles, I can inject it into steam process but obviously the code signature is different than Windows. I got the new pattern for Linux, nvm, I found my way.

I nuked the initial PR, but the code is still here if you're interested https://github.com/m4dEngi/SteamKit/tree/TuxHook2

**Reply by Nemirtingas** [reply]
I did it. It works on my Ubuntu install. Can dump in and outgoing protobufs.
